platform/linux-generic/test/validation/api/ml/ml_linux.c

```diff
@@ -0,0 +1,1259 @@
+	ret = odp_ml_run(model, &data, &run_param);
+	CU_ASSERT_EQUAL(ret, 1);
+
+	/* Test different segment sizes */
+	batch_size = MAX_BATCH_SIZE;
+	odp_ml_run_param_init(&run_param);
+	run_param.result = &result;
+	run_param.batch_size = batch_size;
+	data.input_seg = input_segs;
+	data.output_seg = output_segs;
+
+	for (int seg_size = SIZE; seg_size > 0; seg_size--) {
+		int num_seg = (SIZE + seg_size - 1) / seg_size;
+
+		if ((uint32_t)num_seg > global.ml_capa.max_segs_per_input ||
+		    (uint32_t)num_seg > global.ml_capa.max_segs_per_output)
+			break;
+
+		data.num_input_seg = num_seg * 2;
+		data.num_output_seg = num_seg;
+
+		for (int seg = 0; seg < num_seg; seg++) {
+			int size = seg_size;
+
+			if (seg == num_seg - 1)
+				size = SIZE - seg * seg_size;
+
+			input_segs[seg].addr = (char *)x1 + seg * seg_size;
+			input_segs[seg].size = size;
+			input_segs[seg + num_seg].addr = (char *)x2 + seg * seg_size;
+			input_segs[seg + num_seg].size = size;
+			output_segs[seg].addr = (char *)y + seg * seg_size;
+			output_segs[seg].size = size;
+		}
+
+		memset(y, 0, sizeof(y));
+		ret = odp_ml_run(model, &data, &run_param);
+		CU_ASSERT_EQUAL(ret, 1);
+		if (ret != 1)
+			goto fail;
+
+		for (uint32_t j = 0; j < batch_size * NUM_COLUMN; j++)
+			CU_ASSERT_EQUAL(y[j], y_expected[j]);
+	}
+
+fail:
+	CU_ASSERT_EQUAL_FATAL(odp_ml_model_unload(model, NULL), 0);
+	CU_ASSERT_EQUAL(odp_ml_model_destroy(model), 0);
+}
+
+static void run_global_ml_model(void)
+{
+	int ret = 0;
+	odp_ml_run_result_t result;
+
+	ret = odp_ml_model_load(global.ml_model, NULL);
+	CU_ASSERT_EQUAL_FATAL(ret, 0);
+
+	global.run_param.result = &result;
+
+	ret = odp_ml_run(global.ml_model, &global.data, &global.run_param);
+	CU_ASSERT_EQUAL(ret, 1);
+	CU_ASSERT_FALSE(result.error_code);
+	CU_ASSERT_EQUAL(*(int32_t *)global.output_seg.addr, global.y_expected);
+
+	ret = odp_ml_model_unload(global.ml_model, NULL);
+	CU_ASSERT_EQUAL_FATAL(ret, 0);
+	global.run_param.result = NULL;
+}
+
+static void test_ml_run(void)
+{
+	run_global_ml_model();
+	run_model_batch_add();
+}
+
+static void test_ml_run_multi(void)
+{
+	int ret;
+	int32_t y;
+	int32_t x = 8;
+	int32_t y_expected = 28;
+	odp_ml_data_t data[RUN_NUM];
+	odp_ml_data_seg_t input_seg;
+	odp_ml_data_seg_t output_seg;
+	odp_ml_run_param_t param[RUN_NUM];
+	odp_ml_run_result_t result[RUN_NUM];
+
+	ret = odp_ml_model_load(global.ml_model, NULL);
+	CU_ASSERT_EQUAL_FATAL(ret, 0);
+
+	param[0] = global.run_param;
+	param[0].result = &result[0];
+	param[1].result = &result[1];
+
+	/* Prepare data for running model inference */
+	data[0] = global.data;
+	data[1].num_input_seg = NUM_INPUTS;
+	data[1].input_seg = &input_seg;
+	input_seg.size = sizeof(int32_t);
+	input_seg.addr = &x;
+
+	data[1].num_output_seg = NUM_OUTPUTS;
+	data[1].output_seg = &output_seg;
+	output_seg.size = sizeof(int32_t);
+	output_seg.addr = &y;
+
+	ret = odp_ml_run_multi(global.ml_model, data, param, RUN_NUM);
+	CU_ASSERT_EQUAL(ret, RUN_NUM);
```

Comment:
Implementation could always return one on a multi-call. Test should handle both ret = 2, and ret =1 -> ret =1 as success.

Reply:
Added a loop for this and also for odp_ml_run_start_multi().